enshrine-esdt-safe/src/from_sovereign/transfer_tokens.rs

```diff
@@ -36,13 +39,139 @@ pub trait TransferTokensModule:
+            self.emit_transfer_failed_events(
+                &hash_of_hashes,
+                &OperationTuple {
+                    op_hash: operation_hash,
+                    operation,
+                },
+            );
+
+            sc_panic!("Not enough WEGLD to register all tokens");
+        }
+
+        let minted_operation_tokens = self.mint_tokens(&remaining_tokens);
         let operation_tuple = OperationTuple {
-            op_hash: operation_hash,
-            operation,
+            op_hash: operation_hash.clone(),
+            operation: operation.clone(),
+        };
+
+        self.distribute_payments(
+            hash_of_hashes.clone(),
+            operation_tuple,
+            minted_operation_tokens,
+        );
+    }
+
+    #[endpoint(registerTokens)]
+    #[payable("*")]
+    fn register_tokens(&self, tokens: MultiValueEncoded<TokenIdentifier>) {
+        let call_payment = self.call_value().single_esdt().clone();
+        let wegld_identifier = self.wegld_identifier().get();
+
+        require!(
+            call_payment.token_identifier == wegld_identifier,
+            "WEGLD is the only token accepted as register fee"
+        );
+
+        require!(
+            call_payment.amount == DEFAULT_ISSUE_COST * tokens.len() as u64,
+            "WEGLD fee amount is not met"
+        );
+
+        for token_id in tokens {
+            self.register_token(token_id);
+        }
+    }
+
+    #[endpoint(depositForNewRegistrations)]
+    #[payable("*")]
+    fn deposit_for_new_registrations(&self) {
+        let call_value = self.call_value().single_esdt();
+
+        require!(
+            call_value.amount >= DEFAULT_ISSUE_COST,
+            "The minimum amount of WEGLD wasn't met"
+        );
+    }
+
+    fn get_wegld_payment_info(
+        &self,
+        tokens: &ManagedVec<OperationEsdtPayment<Self::Api>>,
+    ) -> (BigUint, isize) {
+        for (i, token) in tokens.iter().enumerate() {
+            if self.is_wegld(&token.token_identifier) {
+                return (token.token_data.amount, i as isize);
+            }
+        }
+
+        (BigUint::from(0u32), -1)
+    }
+
+    fn verify_operation_tokens_for_issue_fee(
+        &self,
+        sender: &ManagedAddress<Self::Api>,
+        tokens: ManagedVec<OperationEsdtPayment<Self::Api>>,
+    ) -> (ManagedVec<OperationEsdtPayment<Self::Api>>, bool) {
+        require!(!tokens.is_empty(), "Tokens array should not be empty");
+        let (wegld_amount, wegld_position) = self.get_wegld_payment_info(&tokens);
+
+        if wegld_position == -1 {
+            for token in tokens.iter() {
+                if !self.was_token_registered(&token.token_identifier) {
+                    return (ManagedVec::new(), false);
+                }
+            }
+
+            return (tokens, false);
         };
 
-        self.distribute_payments(hash_of_hashes, operation_tuple, minted_operation_tokens);
+        if wegld_position >= 0 && tokens.len() == 1 {
```

Comment:
- `wegld_position >= 0` is always true here, you can delete this condition.
- should return `(tokens, true)`